lib/Sema/TypeCheckAttr.cpp

```diff
@@ -5669,3 +5669,153 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
+               "escaping");
+      hasError |= true;
+    }
+
+    if (handlerTypeAttrs.has(TAK_autoclosure)) {
+      diagnose(handlerTypeAttrs.getLoc(TAK_autoclosure),
+               diag::attr_completion_handler_async_handler_must_not_have_attr,
+               attr, "autoclosure");
+      hasError |= true;
+    }
+
+    if (hasError)
+      return;
+  }
+
+  // Phase 3: Find mapped async function
+  {
+    // Get the list of candidates based on the name
+    // Grab all functions that are async
+    // TODO: Sanity check types -- we just use the DeclName for now
+    //  - Need a throwing decl if the completion handler takes a Result type
+    //    containing an error, or if it takes a tuple containing an optional
+    //    error.
+    //  Find a declref that works.
+    //  Get list of candidates based on the name.
+    //  The correct candidate will need to be async.
+    //
+    //  TODO: Implement the type matching stuff eventually
+    //  If the completion handler takes a single type, then we find the async
+    //  function that returns just that type. (easy case)
+    //
+    //  If the completion handler takes a result type consisting of a type and
+    //  an error, the async function should be throwing and return that type.
+    //  (easy-ish case)
+    //
+    //  If the completion handler takes an optional type and an optional Error
+    //  type, this could map to either of these two. The intent isn't clear.
+    //    - func foo() async throws -> Int
+    //    - func foo() async throws -> Int?
+    //  This case is ambiguous, so we will report an error.
+    //
+    //  If the completion handler takes multiple types, the async function
+    //  should return all of those types in a tuple
+
+    SmallVector<ValueDecl *> allCandidates;
+    lookupReplacedDecl(attr->AsyncFunctionName, attr, attachedFunctionDecl,
+                       allCandidates);
+    SmallVector<AbstractFunctionDecl *> candidates;
+    candidates.reserve(allCandidates.size());
+    for (ValueDecl *candidate : allCandidates) {
+      AbstractFunctionDecl *funcDecl =
+          dyn_cast<AbstractFunctionDecl>(candidate);
+      if (!funcDecl) // Only consider functions
+        continue;
+      if (!funcDecl->hasAsync()) // only consider async functions
+        continue;
+      candidates.push_back(funcDecl);
+    }
+
+    if (candidates.empty()) {
+      diagnose(attr->AsyncFunctionNameLoc,
+               diag::attr_completion_handler_async_no_suitable_function, attr,
+               attr->AsyncFunctionName);
+      return;
+    } else if (candidates.size() > 1) {
+      diagnose(attr->AsyncFunctionNameLoc,
+               diag::attr_completion_handler_async_ambiguous_function, attr,
+               attr->AsyncFunctionName);
+
+      for (AbstractFunctionDecl *candidate : candidates) {
+        diagnose(candidate->getLoc(), diag::decl_declared_here,
+                 candidate->getName());
```

Comment:
Maybe I'm missing something, but I didn't see a way of passing the decl into the diagnose without changing `decl_declared_here` defined in `DiagnosticsSema.def`, which I don't want to do since others are using it. This isn't too bad.

Reply:
That's okay, was just pointing it out since I only recently found it 👍